CMakeLists.txt

```
@@ -177,7 +182,7 @@ if(cached_cmake_build_type)
 endif()
 
 set(default_build_type "RelWithDebInfo")
-if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
+if(NOT CMAKE_BUILD_TYPE AND NOT GENERATOR_IS_MULTI_CONFIG)
```

```suggestion
if(NOT GENERATOR_IS_MULTI_CONFIG)
```
https://cmake.org/cmake/help/latest/prop_gbl/GENERATOR_IS_MULTI_CONFIG.html says "Multi-config generators (...) ignore `CMAKE_BUILD_TYPE`.", so I guess we should do the same?

---

The `NOT GENERATOR_IS_MULTI_CONFIG` condition ensures that we use a single-config generator.

The `NOT CMAKE_BUILD_TYPE` condition checks whether we need to set the default build type. This is actually is the case when the user does not provide `-DCMAKE_BUILD_TYPE=...` on the command line.

---

Going to update this branch shortly with a more readable implementation.

---

[Updated](https://github.com/bitcoin-core/secp256k1/pull/1239#issuecomment-1484070941).